### PR TITLE
refactor: reduce complexity in AlertService.process_all_routes() (#294)

### LINKED ISSUE #294: Refactor alert_service.process_all_routes() to extract pure functional helpers
## Description

The `AlertService.process_all_routes()` method currently has 51 statements, triggering the PLR0915 linting rule (max 50 statements). This method should be refactored to extract pure functional helpers to improve testability, maintainability, and reduce complexity.

## Current State

- Method: `AlertService.process_all_routes()` (app/services/alert_service.py:307)
- Statements: 51 (exceeds limit of 50)
- Suppressed with: `# noqa: PLR0915`

## Proposed Refactoring

Extract pure functions (with no side effects) for:

1. **Schedule validation logic**
   - Check if current time is within route schedule
   - Pure function: takes route, schedule, current time → returns boolean

2. **Disruption filtering logic**
   - Filter disruptions by severity
   - Pure function: takes disruptions, disabled severities → returns filtered disruptions

3. **Route processing decision logic**
   - Determine if route should be processed
   - Pure function: takes route state, disruptions → returns boolean

And any other reasonable extraction.

## Benefits

- **Testability**: Pure functions can be unit tested independently without mocks
- **Maintainability**: Smaller, focused functions are easier to understand and modify
- **Reusability**: Extracted helpers can be reused in other services if needed
- **Complexity**: Reduces cyclomatic complexity of main method

## Implementation Notes

- Follow the pattern in `admin_service.py` where pure calculation functions like `calculate_success_rate()` are extracted to module level
- Keep all extracted functions pure (no side effects, no database/network calls)
- Maintain 100% test coverage on extracted functions
- Simplify integration tests where unit tests provide more efficient test coverage
- Use paramaterised tests for functional helpers where possible to simplify covering all code paths
- Update existing tests to cover the refactored code

## Related

- Part of #290 (OpenTelemetry instrumentation work)
- Follows ADR-12 (Explicit Span Status Pattern)

## Acceptance Criteria

- [ ] `process_all_routes()` has ≤50 statements
- [ ] Extracted helpers are pure functions at module level
- [ ] All tests pass with ≥95% coverage maintained
- [ ] Pre-commit hooks pass (including PLR0915 without noqa)
- [ ] Code is more readable and maintainable

### PR DESCRIPTION
**Pull request by mnbf9rca**
## Summary

Refactored `AlertService.process_all_routes()` to reduce complexity by extracting pure functions and helper methods, improving testability and maintainability. This addresses issue #294.

### 1. Extract Pure Functions (Commit e89d5ff)
Extracted 4 stateless pure functions to module level:
- `get_day_code()` - Convert weekday integer to day code string
- `is_time_in_schedule_window()` - Check if time/day fall within schedule
- `filter_alertable_disruptions()` - Filter disruptions by disabled severities
- `init_alert_processing_stats()` - Initialize stats dictionary

**Benefits:**
- Reduced statement count from 51 to 50
- Removed `# noqa: PLR0915` suppression
- Added 19 comprehensive parameterized unit tests
- Easier to test and reason about

### 2. Extract Helper Methods (Commit 8ac53f6)
Extracted 3 helper methods to reduce nesting:
- `_set_span_stats_attributes()` - Set OTEL span attributes
- `_fetch_global_disruption_data()` - Fetch disruptions and disabled severities
- `_process_single_route()` - Process individual route with error handling

**Benefits:**
- Reduced nesting from 4 levels to 2 levels
- Reduced main method from ~140 lines to ~50 lines
- Better separation of concerns
- Each method has single responsibility

### 3. Add Unit Tests & Simplify Integration Tests (Commit c0d3e20)
Added comprehensive unit tests and removed redundant integration tests:
- Added 13 focused unit tests in 3 test classes
- Removed 3 redundant integration tests (now covered by unit tests)
- Fixed RuntimeWarnings in OTEL tests

**Benefits:**
- Faster test execution
- Better test isolation
- Easier debugging when tests fail
- Clearer test organization

## Test Coverage

- ✅ All 135 tests pass (128 alert_service + 7 OTEL)
- ✅ 96.71% coverage on `alert_service.py` (exceeds 95% requirement)
- ✅ 100% coverage on all new pure functions
- ✅ No RuntimeWarnings

## Metrics

| Metric | Before | After | Improvement |
|--------|--------|-------|-------------|
| Statement count | 51 | ~25-30 | ~50% reduction |
| Nesting levels | 4 | 2 | 50% reduction |
| Main method lines | ~140 | ~50 | 64% reduction |
| Unit tests | 19 pure function tests | 32 total unit tests | +13 focused tests |
| Integration tests | 6 | 3 | Removed redundancy |

## Checklist

- [x] All tests pass
- [x] Code coverage ≥95% on changed code
- [x] Linting passes (ruff)
- [x] Type checking passes (mypy)
- [x] Pre-commit hooks pass
- [x] Follows existing patterns (disruption_helpers.py, admin_service.py)
- [x] No ADR changes required
- [x] Documentation updated (docstrings with examples)

## Related Issues

Closes #294

## Review Notes

The refactoring maintains all existing functionality while significantly improving code quality metrics. The test suite now follows the test pyramid pattern with many fast unit tests and fewer integration tests covering critical end-to-end flows.

## Summary by Sourcery

Refactor alert processing in AlertService to reduce complexity by extracting pure helper functions and dedicated methods while improving observability and test coverage.

Enhancements:
- Extract pure helper functions for day code conversion, schedule-window checks, disruption filtering, and alert stats initialization to simplify alert processing logic.
- Introduce helper methods to fetch global disruption data, set span statistics attributes, and process a single route, significantly reducing nesting and method length in process_all_routes.
- Update alert scheduling logic to reuse the new pure functions, keeping time and day calculations centralized.
- Simplify process_all_routes to orchestrate high-level workflow and always set OTEL span attributes via a dedicated helper for clearer observability.

Tests:
- Add comprehensive unit tests for the new pure functions, span attribute helper, global disruption fetching, and single-route processing behavior across success, error, and edge cases.
- Remove now-redundant process_all_routes integration tests and update OTEL-related tests to target the refactored helpers, eliminating RuntimeWarnings and speeding up the suite.

### CODE REVIEW

File: backend/tests/services/test_alert_otel.py [hunk 1]
```diff
@@ -125,47 +111,22 @@ async def test_process_all_routes_mixed_success_and_failure(
...
-            mock_execute_result = AsyncMock()
-            mock_scalars = MagicMock()
-            mock_scalars.all.return_value = []
-            mock_execute_result.scalars.return_value = mock_scalars
-            mock_db.execute.return_value = mock_execute_result
-
-            # Act
-            stats = await alert_svc.process_all_routes()
+                return 0, True  # No alerts sent, error occurred
+            return 0, False
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/services/test_alert_otel.py:124]: Thank you for the suggestion, but I cannot accept this change. The suggested ternary expression drops the first case that handles `successful_route`, which returns `(3, False)` to indicate 3 alerts were sent successfully.

The test requires this behavior to verify the assertion `assert stats["alerts_sent"] == 3` on line 133. The original code handles three cases:
1. `successful_route` → `(3, False)` - 3 alerts sent
2. `failing_route` → `(0, True)` - error occurred  
3. Any other route → `(0, False)` - default

The suggested ternary only handles cases 2 and 3, which would break the test.

File: backend/app/services/alert_service.py [hunk 2]
```diff
@@ -304,152 +454,69 @@ async def _log_line_disruption_state_changes(
...
-                    logger.error(
-                        "disruption_logging_failed",
-                        error=str(e),
-                        exc_info=e,
-                    )
-                    # Don't track as error - per-route processing will handle its own disruptions
-                    # This is just for centralized logging
+                # Fetch global disruption data once for all routes
+                # Errors are non-fatal - returns empty data on failure
+                _, disabled_severity_pairs = await self._fetch_global_disruption_data()
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/alert_service.py]: Good catch! Simplified the API in commit 7412658.

Changes:
- Updated `_fetch_global_disruption_data()` return type to `set[tuple[str, int]]`
- Method still fetches disruptions internally for `_log_line_disruption_state_changes()` but only returns the disabled severity pairs that callers actually need
- Updated call site to not use tuple unpacking
- Updated test mocks in both test files

All 135 tests pass with 96.72% coverage on alert_service.py.

**Review by mnbf9rca — COMMENTED** [review]
Thank you for the detailed review! I've addressed all the comments in commit 7412658:

**✅ AsyncMock in TestProcessSingleRoute** - Fixed all async method mocks to use `new_callable=AsyncMock` for `_get_active_schedule`, `_get_route_disruptions`, `_should_send_alert`, and `_send_alerts_for_route`. More correct and robust async testing pattern.

**✅ Unused return value** - Simplified `_fetch_global_disruption_data()` to return only `set[tuple[str, int]]`. The method internally uses disruptions for logging via `_log_line_disruption_state_changes()`, but the caller only needs disabled severity pairs. Cleaner API following YAGNI.

**✅ Lost per-disruption logging** (modified) - Added `filtered_count` to the existing log entry instead of verbose per-item logging. This provides debug visibility in a single log entry while keeping the pure function design intact. The function remains pure (no side effects) and the call site provides the observability.

All 135 tests pass, 96.72% coverage on alert_service.py, type checking and linting pass.
